File: 2/javascript/problem2.js
var Generator = require('./Generator.js').Generator;

exports.answer = function () {
  return fibs()
    .takeWhile(function (x) { return x < 4000000; })
    .filter(function (x) { return x % 2 === 0; })
    .sum();
};

function fibs() {
  var x = [ 0, 1 ];
  return new Generator(function() {
    x = [ x[1], x[0]+x[1] ];
    return x[0];
  });
}

File: 2/haskell/src/Problem2.hs
module Problem2 where

answer :: Integer
answer = fibs
  # takeWhile(< 4000000)
  # filter(\x -> x `mod` 2 == 0)
  # sum

fibs :: [Integer]
fibs = 0 : 1 : zipWith (+) fibs (tail fibs)

infixl 8 #
(#) :: a -> (a -> b) -> b
a # f = f a
